### PR TITLE
Inject git information into clickhouse binary file

### PR DESCRIPTION
**Pull request by taiyang-li**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Inject git information into clickhouse binary file. So we can get source code revision easily from clickhouse binary file.
 
Detailed description / Documentation draft:
``` 
$ ./clickhouse server --version                
ClickHouse server version 21.13.1.1.
GIT SHA1: 81e2e72dd07120708ce3b4fdd3886e7884fe84ed
GIT BRANCH: inject_gitinfo
GIT DATE: Thu Dec 23 20:52:36 2021
GIT COMMIT SUBJECT: fix bug

$ ./clickhouse client --version                           
ClickHouse client version 21.13.1.1.
GIT SHA1: 81e2e72dd07120708ce3b4fdd3886e7884fe84ed
GIT BRANCH: inject_gitinfo
GIT DATE: Thu Dec 23 20:52:36 2021
GIT COMMIT SUBJECT: fix bug

$ ./clickhouse local --version 
ClickHouse local version 21.13.1.1.
GIT SHA1: 81e2e72dd07120708ce3b4fdd3886e7884fe84ed
GIT BRANCH: inject_gitinfo
GIT DATE: Thu Dec 23 20:52:36 2021
GIT COMMIT SUBJECT: fix bug
```

### CODE REVIEW

**Comment by taiyang-li** [comment]
Failed tests seems irrelevant to this PR.

**Comment by alesapin** [comment]
`01526_client_start_and_exit` probably relevant.

**Comment by alexey-milovidov** [comment]
Maybe put this info into `system.build_options`? 
We already have git hash there.

**Comment by taiyang-li** [comment]
> Maybe put this info into `system.build_options`? We already have git hash there.
ok.  I will try.

**Comment by taiyang-li** [comment]
@alexey-milovidov the git info I inject into binary has a little different from those in `system.build_options`.  The former update git info after each building of clickhouse binary. But the latter only update after each release.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -20,13 +20,41 @@ include(../cmake/limit_jobs.cmake)
...
 set (CONFIG_COMMON "${CMAKE_CURRENT_BINARY_DIR}/Common/config.h")
+set (GIT_INFO "${CMAKE_CURRENT_BINARY_DIR}/Common/git_info.h")
 
 include (../cmake/version.cmake)
 message (STATUS "Will build ${VERSION_FULL} revision ${VERSION_REVISION} ${VERSION_OFFICIAL}")
 configure_file (Common/config.h.in ${CONFIG_COMMON})
 configure_file (Common/config_version.h.in ${CONFIG_VERSION})
 configure_file (Core/config_core.h.in "${CMAKE_CURRENT_BINARY_DIR}/Core/include/config_core.h")
 
+set(GIT_EXECUTABLE /usr/bin/git)
```
Comment (qoega) [inline src/CMakeLists.txt:31]: Probably we need a case when git is unavailable or we build not from repo.

**Comment by alexey-milovidov** [comment]
> Should I add a new key named GITHASH to represent the git hash when building, or just override the old key VERSION_GITHASH ?

We can add new entries. So, we will have two git hashes, and they will coincide for release builds.

File: src/CMakeLists.txt (hunk 1 above)
Comment (taiyang-li) [reply src/CMakeLists.txt:31]: ok.

**Comment by taiyang-li** [comment]
@alexey-milovidov   ok

**Comment by taiyang-li** [comment]
```  
SELECT *
FROM system.build_options
WHERE name LIKE 'GIT%'

Query id: dbdd5874-a0b4-48f9-a743-b612dc61bbab

┌─name───────────────┬─value────────────────────────────────────┐
│ GIT_HASH           │ a88b2975c0589e24b1fa705de2d0818cf3ddad5d │
│ GIT_BRANCH         │ inject_gitinfo                           │
│ GIT_DATE           │ Sat Dec 25 15:45:50 2021                 │
│ GIT_COMMIT_SUBJECT │ inject into system.build_options         │
└────────────────────┴──────────────────────────────────────────┘
```

**Comment by alexey-milovidov** [comment]
We should have date in ISO-8601 format, in UTC.

**Comment by taiyang-li** [comment]
> We should have date in ISO-8601 format, in UTC. 
ok. I will fix it

**Comment by taiyang-li** [comment]
```
SELECT *
FROM system.build_options
WHERE name LIKE 'GIT%'

Query id: 5ae23051-5e54-4afb-84f1-33eb7873b338

┌─name───────────────┬─value────────────────────────────────────┐
│ GIT_HASH           │ a88b2975c0589e24b1fa705de2d0818cf3ddad5d │
│ GIT_BRANCH         │ inject_gitinfo                           │
│ GIT_DATE           │ 2021-12-25 07:45:50 +0000                │
│ GIT_COMMIT_SUBJECT │ inject into system.build_options         │
└────────────────────┴──────────────────────────────────────────┘

4 rows in set. Elapsed: 0.002 sec. 

```

**Comment by taiyang-li** [comment]
> `01526_client_start_and_exit` probably relevant.

The error didn't see again.

**Review by qoega — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
This PR has a bug: the string is not escaped.

File: src/Storages/System/StorageSystemBuildOptions.generated.cpp.in [hunk 2]
```diff
@@ -50,6 +50,10 @@ const char * auto_config_build[]
     "USE_KRB5", "@USE_KRB5@",
     "USE_FILELOG", "@USE_FILELOG@",
     "USE_BZIP2", "@USE_BZIP2@",
+    "GIT_HASH", "@GIT_HASH@",
+    "GIT_BRANCH", "@GIT_BRANCH@",
+    "GIT_DATE", "@GIT_DATE@",
+    "GIT_COMMIT_SUBJECT", "@GIT_COMMIT_SUBJECT@",
```
Comment (alexey-milovidov) [inline src/Storages/System/StorageSystemBuildOptions.generated.cpp.in:56]: Here is the bug.
GIT_COMMIT_SUBJECT can contain quotes `"`, backslashes `\` and possibly even newlines.

File: src/Storages/System/StorageSystemBuildOptions.generated.cpp.in [hunk 3]
```diff
@@ -50,6 +50,10 @@ const char * auto_config_build[]
     "USE_KRB5", "@USE_KRB5@",
     "USE_FILELOG", "@USE_FILELOG@",
     "USE_BZIP2", "@USE_BZIP2@",
+    "GIT_HASH", "@GIT_HASH@",
+    "GIT_BRANCH", "@GIT_BRANCH@",
```
Comment (alexey-milovidov) [inline src/Storages/System/StorageSystemBuildOptions.generated.cpp.in:54]: I will be not surprized if git branch name can also contain something unusual.

**Comment by alexey-milovidov** [comment]
Fixed in ff779b6409.

**Pull request merged**